kernel/src/process.rs

```diff
@@ -1143,43 +1289,174 @@ impl<C: Chip> ProcessType for Process<'_, C> {
-            return Err(ReturnCode::FAIL);
+            // Do not operate on an inactive process
+            return SyscallReturn::AllowReadOnlyFailure(ErrorCode::FAIL, buf_start_addr, size);
         }
 
-        match NonNull::new(buf_start_addr as *mut u8) {
-            None => {
-                // A null buffer means pass in `None` to the capsule
-                Ok(None)
+        // A process is allowed to pass any pointer if the slice
+        // length is 0, as to revoke kernel access to a memory region
+        // without granting access to another one
+        let new_slice = if size == 0 {
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as a slice
+            // of length 0 will never allow dereferencing any memory
+            // in a safe manner.
+            unsafe { ReadOnlyAppSlice::new(buf_start_addr, 0, self.appid()) }
+        } else if self.in_app_owned_memory(buf_start_addr, size)
+            || self.in_app_flash_memory(buf_start_addr, size)
+        {
+            // TODO: Check for buffer aliasing here
+
+            // Valid slice, we need to adjust the app's watermark
+            // note: in_app_owned_memory ensures this offset does not wrap
+            let buf_end_addr = buf_start_addr.wrapping_add(size);
+            let new_water_mark = cmp::max(self.allow_high_water_mark.get(), buf_end_addr);
+            self.allow_high_water_mark.set(new_water_mark);
+
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as long as
+            // we make sure that we're pointing towards userspace
+            // memory (verified using `in_app_owned_memory` or
+            // `in_app_flash_memory`) and respect alignment and other
+            // constraints of the Rust references created by
+            // ReadWriteAppSlice.
+            unsafe { ReadOnlyAppSlice::new(buf_start_addr, size, self.appid()) }
+        } else {
+            return SyscallReturn::AllowReadOnlyFailure(ErrorCode::INVAL, buf_start_addr, size);
+        };
+
+        let allow_result = driver_invoc_closure(new_slice);
```

Comment:
Trying to pick this up. This abstraction was introduced in 

https://github.com/tock/tock/commit/50743b25f37080936ecd0ffb82140a039347b925

I believe the principal reason it's a closure is the concern that correctly enforcing swapping semantics at runtime might require per-process state which is modified by this operation. For example, if the process keeps a table of what buffers are currently allowed, then when a swap occurs it needs to update this table (remove the old entry, add the new one). Otherwise, you have the scheduler explicitly performing this swap in the process tables after it invokes the driver, which would require the scheduler to have access to the table. Encapsulating this as part of the call stack (`ProcessType` returns after it updates its tables) seems much simpler and less prone to bugs.

Reply:
Using a closure has a similar effect, while updating any process state is in process.rs, the process is now also implementing the scheduler. The code for how app slices are handled and that implements the spec is not in sched entirely, and requires looking at the process implementation. That makes it more difficult to re-implement `Process` (as has been done for host-side testing), and I think is an interesting future opportunity as well. By putting the logic in sched, we can add APIs to Process that make it clear what complexity a Process implementation needs to handle.